### PR TITLE
Add first-class Codex provider

### PR DESCRIPTION
**Pull request by Logo-web**
## Summary

This adds a first-class `codex` provider to Kong, backed by the ChatGPT Codex OAuth flow instead of the standard OpenAI API.

## What changed

- added `LLMProvider.CODEX`
- added `CodexClient` for ChatGPT Codex backend requests
- added `codex_auth` helpers for loading and refreshing local Codex OAuth credentials
- wired Codex into `kong setup`, provider resolution, endpoint probing, and CLI messaging
- kept existing `openai` and `custom` provider behavior intact
- added tests for Codex auth, Codex client flow, setup, probe, and provider routing
- updated README usage docs

## Why

Kong previously could reuse a local Codex login only as a credential source for the normal OpenAI API path. That still fails for accounts without OpenAI API billing.

This change adds a real Codex-backed path so users can run Kong through their local ChatGPT/Codex login directly.

## Verification

- `uv run pytest -q`
- full suite passing: `448 passed`
- local Codex probe verified successfully

---
## Summary by cubic
Adds a first-class `codex` provider that talks to the ChatGPT Codex backend via OAuth, independent of the OpenAI API. Users can run Kong with a local ChatGPT login; setup, probing, routing, and usage reporting were updated (tokens only, no cost).

- New Features
  - Added `LLMProvider.CODEX` and a Codex client with SSE streaming, tool-calling, token usage, and OAuth auto-refresh on 401s.
  - Auth: `codex_auth` loads/refreshes from `~/.codex/auth.json` (override with `CODEX_HOME`); `openai_auth` now auto-resolves `OPENAI_API_KEY`, Codex API key, or Codex OAuth; the OpenAI client re-resolves credentials dynamically.
  - Setup/UX: wizard adds `codex` and `Anthropic + Codex`, shows credential source labels, and guides to run `codex`; provider labels/help updated; default Codex model is `gpt-5-codex`; clearer errors when Codex login is missing or when OpenAI uses Codex OAuth without API billing.
  - Probing: Codex probe added; OpenAI probe falls back to an inference check if model-list scope is missing.
  - Usage/Export/Docs: treat `codex` as non-billable (token counts only; exports mark cost tracking off); README adds Codex sign-in (`codex`), `CODEX_HOME`, separate billing notes, and `--provider codex` examples.

- Migration
  - No breaking changes; `openai` and `custom` behavior unchanged.
  - To use Codex: run `codex` to sign in, then `kong setup` and choose `codex` (or `Anthropic + Codex`).
  - If `openai` runs via Codex OAuth and requests fail, enable OpenAI API billing or set `OPENAI_API_KEY`.

<sup>Written for commit 300ad957eeb302fb46d13e1300b6418b53b786ff. Summary will update on new commits.</sup>

### CODE REVIEW

**Review by amruth-sn — DISMISSED** [review]
Generally LGTM. Thanks for the issue and I appreciate the fixes!

**Review by amruth-sn — COMMENTED** [review]
## Comprehensive PR Review

Reviewed with 5 specialized agents: code-reviewer, silent-failure-hunter, type-design-analyzer, pr-test-analyzer, comment-analyzer.

**Summary:** 4 HIGH, 8 MEDIUM, 7 LOW findings across 18 files. The PR is well-structured overall — clean integration with the multi-provider architecture, thorough setup wizard tests, and good credential resolution chain. The main concerns are around silent error handling in the auth layer and some type/convention issues.

### Test Coverage Gaps (LOW)

The following test gaps were identified:

1. **401 retry with token refresh is untested** — The `_request` method's 401 → refresh → retry logic is the primary resilience mechanism for expired tokens. Missing tests: successful retry after refresh, refresh returns `None`, double 401 failure.

2. **SSE stream with no `response.completed` event is untested** — A truncated or errored stream that never sends `response.completed` hits a specific error path (`CodexClientError("Codex backend did not return a completed response.")`) with no coverage.

3. **`MagicMock` used for `CodexCredential` in client tests** — `test_codex_client.py` uses `MagicMock(access_token="oauth", account_id="acct", refresh_token="refresh")` instead of real `CodexCredential` instances. This means tests pass even if `CodexCredential` gains a required field or changes attribute names.

4. **SSE parser has no isolated unit tests** — `_read_sse_events` and `_append_buffered_event` are a non-trivial state machine tested only through integration. Missing: multi-line data events, interleaved event types, non-JSON payloads, malformed auth file edge cases in `resolve_codex_credential`.

File: kong/llm/codex_auth.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        "client_id": CLIENT_ID,
+    }).encode("utf-8")
+    request = urllib.request.Request(
+        TOKEN_URL,
+        data=payload,
+        method="POST",
+        headers={"Content-Type": "application/x-www-form-urlencoded"},
+    )
+
+    with urllib.request.urlopen(request, timeout=30) as response:
```
Comment (amruth-sn) [inline kong/llm/codex_auth.py:70]: **HIGH: Unhandled exception in `urlopen` crashes during 401 retry**

`urllib.request.urlopen` can raise `HTTPError` (e.g. 400 for invalid/revoked refresh token), `URLError` (network unreachable), or `socket.timeout`. None are caught here.

The caller in `CodexClient._request` calls this inside an `except urllib.error.HTTPError` block — Python does not re-enter an active `except` handler for new exceptions of the same type. So a transient token endpoint failure produces a raw `urllib` traceback instead of a clean `CodexClientError`.

```python
try:
    with urllib.request.urlopen(request, timeout=30) as response:
        data = json.loads(response.read().decode("utf-8"))
except (urllib.error.HTTPError, urllib.error.URLError, json.JSONDecodeError, OSError) as exc:
    logger.warning("Codex token refresh failed: %s", exc)
    return None
```

File: kong/llm/codex_auth.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    if not isinstance(parsed, dict):
+        return None
+    return parsed
+
+
+def _restrict_auth_file_permissions(path: Path) -> None:
+    try:
+        os.chmod(path, 0o600)
+    except OSError:
```
Comment (amruth-sn) [inline kong/llm/codex_auth.py:174]: **HIGH: Security — `except OSError: pass` silently leaves credentials world-readable**

This function protects OAuth credentials (access token + refresh token) from being world-readable. If `chmod` fails (permission issues, read-only FS, SELinux/AppArmor), the auth file remains with its default permissions — potentially `0o644` — with zero indication to the user.

```python
def _restrict_auth_file_permissions(path: Path) -> None:
    try:
        os.chmod(path, 0o600)
    except OSError:
        logger.warning(
            "Could not set restrictive permissions on %s — "
            "credentials may be readable by other users",
            path,
            exc_info=True,
        )
```

File: kong/llm/codex_auth.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+
+def resolve_codex_credential() -> CodexCredential | None:
+    auth_path = codex_auth_path()
+    if not auth_path.is_file():
+        return None
+
+    try:
+        auth = json.loads(auth_path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError):
```
Comment (amruth-sn) [inline kong/llm/codex_auth.py:47]: **HIGH: Silent swallow makes corrupted auth file indistinguishable from missing file**

This `except` catches both `OSError` (permission denied, disk I/O errors, symlink loops) and `json.JSONDecodeError` (malformed/truncated auth files). In all cases it returns `None`, which propagates as "no credential found."

A user who has legitimately signed in via `codex` but whose `auth.json` gets corrupted (e.g. crash during `refresh_codex_credential` write) will see "Run `codex` to sign in" — even though they already signed in. They'll keep re-authenticating with no indication the file is corrupted.

Suggestion: split into "file not found" (already handled above) vs. "file exists but is unreadable/corrupt" (log a warning):

```python
try:
    raw = auth_path.read_text(encoding="utf-8")
except OSError:
    logger.warning("Could not read Codex auth file at %s", auth_path, exc_info=True)
    return None
try:
    auth = json.loads(raw)
except json.JSONDecodeError:
    logger.warning("Codex auth file at %s is not valid JSON", auth_path)
    return None
```

File: kong/llm/codex_auth.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        return None
+
+    return _credential_from_auth(auth, auth_path=auth_path)
+
+
+def refresh_codex_credential(credential: CodexCredential | None = None) -> CodexCredential | None:
```
Comment (amruth-sn) [inline kong/llm/codex_auth.py:53]: **HIGH: 5 silent `return None` paths with no logging in `refresh_codex_credential`**

This function returns `None` for 5+ distinct failure modes:
1. No credential / no refresh token (line 56)
2. `urlopen` network failures (unhandled — see other comment)
3. `access_token` missing from server response (line 75)
4. `refresh_token` missing from server response (line 77)
5. `account_id` extraction fails (line 81)

All propagate as "refresh failed" with zero diagnostics. The server might return `{"error": "invalid_grant", "error_description": "Refresh token expired"}` — that's silently discarded because the response lacks `access_token`.

When token refresh fails, every subsequent request fails with 401, and the user has no idea why.

Suggestion: add `logger.warning(...)` or `logger.debug(...)` at each `return None` path, especially cases 3-5 where the server *did* respond but with unexpected content.

File: kong/llm/codex_auth.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+from typing import Any
+
+
+CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
+TOKEN_URL = "https://auth.openai.com/oauth/token"
+JWT_CLAIM_PATH = "https://api.openai.com/auth"
+
+
+@dataclass(frozen=True)
+class CodexCredential:
```
Comment (amruth-sn) [inline kong/llm/codex_auth.py:22]: **MEDIUM: `CodexCredential` needs `__post_init__` validation**

The dataclass accepts empty strings for `access_token` and `account_id`. Invariants (non-empty required fields) are enforced only by factory functions (`_credential_from_auth`, `refresh_codex_credential`), not by the type itself. Anyone can construct `CodexCredential(access_token="", account_id="")` and get a formally valid instance that immediately causes a 401.

```python
def __post_init__(self) -> None:
    if not self.access_token:
        raise ValueError("access_token must be non-empty")
    if not self.account_id:
        raise ValueError("account_id must be non-empty")
```

Same applies to `OpenAICredential.token` in `openai_auth.py`.

File: kong/llm/openai_auth.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    @property
+    def masked_value(self) -> str | None:
+        if self.source == "codex_oauth":
+            return None
+        if len(self.token) <= 11:
+            return "***"
+        return self.token[:7] + "..." + self.token[-4:]
+
+
+def codex_auth_path() -> Path:
```
Comment (amruth-sn) [inline kong/llm/openai_auth.py:37]: **MEDIUM: Duplicated `codex_auth_path()` — import from `codex_auth.py` instead**

This function is identical to `codex_auth_path()` in `kong/llm/codex_auth.py` (same `CODEX_HOME` env var, same `~/.codex/auth.json` fallback). If the path logic changes, both must be updated.

```python
from kong.llm.codex_auth import codex_auth_path
```

File: kong/llm/codex_client.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+                "type": "function",
+                "name": tool["name"],
+                "description": tool["description"],
+                "parameters": tool["input_schema"],
+                "strict": True,
+            })
+        return converted
+
+    @staticmethod
+    def _read_sse_events(response: Any) -> list[tuple[str, dict[str, Any]]]:
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:290]: **MEDIUM: `response: Any` violates CLAUDE.md "Try not to use Any"**

`urllib.request.urlopen` returns `http.client.HTTPResponse`, which has the `.readline()` method being called here. This is the most avoidable `Any` in the PR — a well-known stdlib type.

More broadly, there are ~15 `Any` usages across this file. Some are justified (decoded JSON), but several have known shapes:
- `_user_message` always returns `{"type": "message", "role": "user", "content": [...]}`
- `_convert_tools` has a known output shape
- `_CodexEnvelope.response` and `.outputs` could use `TypedDict`

Suggestion: at minimum, type this parameter as `http.client.HTTPResponse` and consider `TypedDict` for the fixed-shape dicts.

File: kong/llm/codex_client.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+                result.input_tokens = total_input
+                result.output_tokens = total_output
+                result.raw = last_text
+                return result
+
+            for tool_call in tool_calls:
+                raw_args = tool_call.get("arguments", "") or "{}"
+                try:
+                    arguments = json.loads(raw_args)
+                except json.JSONDecodeError:
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:114]: **MEDIUM: Malformed tool-call arguments silently default to empty dict**

When the LLM returns invalid JSON in tool call arguments, this replaces them with `{}` and proceeds to execute the tool with no arguments. The tool will fail with a confusing `KeyError` or `TypeError` about missing required params — with no indication the root cause was malformed LLM output.

```python
except json.JSONDecodeError:
    logger.warning(
        "Tool call %r returned unparseable arguments: %s",
        tool_call.get("name"),
        raw_args[:200],
    )
    arguments = {}
```

Or better — skip the tool call and feed an error result back to the model.

File: kong/llm/codex_client.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+            if not line:
+                CodexClient._append_buffered_event(events, event_name, data_lines)
+                event_name = ""
+                data_lines = []
+                continue
+
+            if line.startswith("event:"):
+                event_name = line.split(":", 1)[1].strip()
+            elif line.startswith("data:"):
+                data_lines.append(line.split(":", 1)[1].lstrip())
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:311]: **MEDIUM: `.lstrip()` strips all leading whitespace instead of exactly one space per SSE spec**

The [SSE specification](https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation) says: "If value is not empty, and value's first character is a U+0020 SPACE character, then remove it from value." Only a single leading space should be stripped.

`.lstrip()` strips all leading whitespace. Not a practical bug today (JSON is whitespace-insensitive), but deviates from spec behavior.

```python
raw_value = line.split(":", 1)[1]
data_lines.append(raw_value[1:] if raw_value.startswith(" ") else raw_value)
```

File: kong/config.py [hunk 10]
```diff
@@ -12,13 +12,15 @@
 class LLMProvider(Enum):
     ANTHROPIC = "anthropic"
     OPENAI = "openai"
+    CODEX = "codex"
```
Comment (amruth-sn) [inline kong/config.py:15]: **MEDIUM: Scattered non-billable checks — add `is_billable` property to `LLMProvider`**

The pattern `provider in (LLMProvider.CUSTOM, LLMProvider.CODEX)` is now repeated in:
- `kong/__main__.py` (`_print_final_stats`, `analyze`)
- `kong/export/structured.py` (`_build_stats_section`)
- `kong/llm/usage.py` (`get_pricing`)

If a fifth provider is added, all must be updated. Consider:

```python
@property
def is_billable(self) -> bool:
    return self not in (LLMProvider.CUSTOM, LLMProvider.CODEX)
```

Also: `display_name` uses a dict lookup that isn't exhaustiveness-checked. If a variant is added without updating the dict, it raises `KeyError` at runtime.

File: kong/llm/openai_auth.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    if env_key:
+        return OpenAICredential(token=env_key, source="env")
+
+    auth_path = codex_auth_path()
+    if not auth_path.is_file():
+        return None
+
+    try:
+        auth = json.loads(auth_path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError):
```
Comment (amruth-sn) [inline kong/llm/openai_auth.py:55]: **MEDIUM: Same silent `(OSError, json.JSONDecodeError) → None` pattern as `codex_auth.py`**

A corrupted or permission-denied auth file is indistinguishable from "no auth file exists." Since this function resolves OpenAI credentials with Codex OAuth as a fallback, a corrupted file silently prevents the fallback from working.

User sees "OPENAI_API_KEY not set" with no mention of the corrupted Codex auth file.

Same fix as `codex_auth.py` — add a logger and distinguish "file not found" from "file exists but is unreadable." (See comment on `codex_auth.py:47`.)

File: kong/llm/probe.py [hunk 12]
```diff
@@ -47,10 +52,21 @@ def _probe_custom(config: LLMConfig) -> bool:
...
+        api_key = credential.token
+
     try:
-        client = openai.OpenAI(api_key=config.api_key, base_url=config.base_url)
+        client = openai.OpenAI(api_key=api_key, base_url=config.base_url)
         client.models.list()
         return True
+    except openai.PermissionDeniedError as e:
+        message = str(e)
+        if "api.model.read" not in message:
```
Comment (amruth-sn) [inline kong/llm/probe.py:66]: **MEDIUM: Fragile error-message substring matching for probe fallback**

`"api.model.read" not in message` checks for a specific scope name inside an exception message string. If OpenAI changes the error message wording, this fallback silently stops working — valid Codex OAuth credentials get rejected with "Could not connect to LLM endpoint."

Suggestion: at minimum, add a comment explaining the heuristic and log when the fallback triggers:

```python
except openai.PermissionDeniedError as e:
    message = str(e)
    # Codex OAuth tokens may lack api.model.read scope.
    # Fall back to inference probe. If OpenAI changes this
    # error message, this heuristic needs updating.
    if "api.model.read" not in message:
        logger.warning("OpenAI credential was rejected: %s", message)
        return False
    logger.debug("models.list() denied, falling back to inference probe")
    return _probe_openai_inference(config, api_key)
```

File: kong/llm/codex_client.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+DEFAULT_MODEL = "gpt-5-codex"
+CODEX_RESPONSES_URL = "https://chatgpt.com/backend-api/codex/responses"
+
+
+class CodexClientError(RuntimeError):
+    """Raised when the Codex backend rejects a request."""
+
+
+@dataclass
+class _CodexEnvelope:
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:29]: **LOW: `_CodexEnvelope` should be `frozen=True` and reduce `Any`**

This dataclass is never mutated after construction in `_perform_request`. Making it frozen would match `CodexCredential`'s immutability pattern.

`response: dict[str, Any]` and `outputs: list[dict[str, Any]]` are the heaviest `Any` usage in the PR. If the response has a known shape (which it does — Codex Responses API schema), consider a lightweight `TypedDict` for at least the top-level keys.

File: kong/llm/codex_client.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+        data = "\n".join(data_lines)
+        if data == "[DONE]":
+            return
+        try:
+            events.append((event_name, json.loads(data)))
+        except json.JSONDecodeError:
+            logger.debug("Skipping non-JSON SSE payload: %s", data[:200])
+
+    def _record_usage(self, model: str, input_tokens: int, output_tokens: int) -> None:
+        usage = self.usage._get(model)
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:332]: **LOW: `self.usage._get(model)` accesses a nominally private method**

`TokenUsage._get` is called by all three client implementations (`AnthropicClient`, `OpenAIClient`, `CodexClient`). It's effectively a public API with a misleading name.

Consider renaming to `get_or_create` or adding a public `record(model, input_tokens, output_tokens)` method to `TokenUsage`.

File: kong/llm/codex_client.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+            CODEX_RESPONSES_URL,
+            data=json.dumps(payload).encode("utf-8"),
+            method="POST",
+        )
+        for key, value in {
+            "Authorization": f"Bearer {credential.access_token}",
+            "chatgpt-account-id": credential.account_id,
+            "OpenAI-Beta": "responses=experimental",
+            "originator": "codex_cli_rs",
+            "Content-Type": "application/json",
```
Comment (amruth-sn) [inline kong/llm/codex_client.py:219]: **LOW: Missing "why" comments on non-obvious architectural decisions**

Several choices in this module are non-obvious and will confuse future maintainers:

1. **Why raw `urllib` instead of the OpenAI SDK?** The Codex Responses endpoint at `chatgpt.com` is not part of the `api.openai.com` surface and requires headers (`chatgpt-account-id`, `originator`) the SDK doesn't support.

2. **`"originator": "codex_cli_rs"`** — This mimics the Rust Codex CLI's identity string. If the backend keys behavior off this value, that has implications.

3. **`CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"`** (in `codex_auth.py`) — Where does this come from? Is it the official Codex CLI's OAuth client ID?

4. **`"include": ["reasoning.encrypted_content"]`** (line 200) — No comment explains why encrypted reasoning content is requested or whether it's consumed anywhere.

These are exactly the kind of rare "why" comments CLAUDE.md says should exist.
